lib/broadway/options.ex

```diff
@@ -218,10 +218,28 @@ defmodule Broadway.Options do
               """
             ],
             batch_size: [
-              type: :pos_integer,
+              type: {:custom, __MODULE__, :validate_batch_size, []},
               default: 100,
               doc: """
-              The size of the generated batches.
+              The size of the generated batches. Default value is `100`.
+              It can also be tuple of {:fun/2, init_size}.  The function must have an arity of 2
+              with `Broadway.Message` and last calculated `acc` returned from it.  `init_size` is provided
+              as the second parameter on first call.
+              The function must return either {:emit, acc} to indicate a full batch,
+              or {:cont, acc} to continue.
+              Default batch size is actually converted to this:
+              `{fn _message, remained ->
+                 remained = remained - 1
+                 if(remained == 0, do: {:emit, batch_size}, else: {:cont, remained})
+               end, batch_size}`
+              """
+            ],
+            max_demand: [
+              type: :pos_integer,
+              doc: """
+              Set the maximum demand of batcher stages. No default value is given because
+              max_demand will take `:batch_size` if this is not given for compatibility reason.
+              Must be set if the `:batch_size` is a function.
```

Comment:
```suggestion
              Set the maximum demand of batcher stages. No default value is given because
              max_demand will take `:batch_size` if this is not given for compatibility reason.
              Must be set if the `:batch_size` is a function.
```
```suggestion
              Sets the maximum demand of batcher stages.

              By default it is set to `:batch_size`, if `:batch_size` is an integer.
              Must be set if the `:batch_size` is a function.
```